File: ams/uttps/scope.py
import numpy as np
from math import sqrt

f = open('d', 'rb')
data = f.read()
f.close()

print("len(data) = ", str(len(data)))

data_us = []
data_us_len = int(len(data)/2.)
k = 0
for i in range(data_us_len): 
    j = i*2 # index of unsigned short array
    d_us = (0x00FF & data[j] | ( (0x00FF & data[j+1]) << 8) )
    if (d_us & 0x8000 != 0) and (k == 0):
        k = i

    data_us.append(d_us)
    i+=2 # step 2, over bytes array

print('len(data_us) = {}'.format(data_us_len))
print('data[{:2d}] = {:#x}'.format(k, data_us[k]))

m = 18 # number of sectors 
turns = data_us[k:] 
nturns = int(len(turns)/m -1) # why -1? 
print('nturns = {}'.format(nturns))
print('len(turns) = {}'.format(len(turns)))
print('2273 * 18 = {}'.format(2273 * 18))

nsectors = m * nturns # total number of sectors
sum1 = np.zeros(m) # sum of sector_values over all turns
sum2 = np.zeros(m) # sum of sector_v_squares over all turns
print(nturns)

for j in range(nturns):
    turn = turns[ j*m : (j+1)*m ] # 18 sectors of each turn
    for i in range(m):
        sector_value = turn[i]
        v = ( (0x7FFF & sector_value) +0.5 ) /5.0
        sum1[i] += v
        sum2[i] += v * v

sectorM = np.sum(sum1) / nsectors
Srpm = 60. * 1000000. / (m * sectorM)
print('Nturns={:d} RPM={:7.1f} sec ava={:6.2f} nsectors={:8d}\n'.format(
    nturns,
    Srpm,
    sectorM,
    nsectors))

for i in range(m):
    s_i = sum1[i]/nturns
    m_i = sectorM - s_i
    v_i = 100. * m_i / sectorM
    x = (sum2[i] - pow(sum1[i],2)/nturns ) / nturns
    sigma_i = sqrt( x )
    print( 'sec {a0:2d}  {a1:7.2f} usec  {a2:7.2f}  {a3:5.1f}   {a4:5.2f}'.format(
        a0 = i,
        a1 = s_i,
        a2 = m_i,
        a3 = v_i,
        a4 = sigma_i
    ))

File: ams/uttps/scope.c
//  scope.c

#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h>
#include <ctype.h>
#include <signal.h>
#include <time.h>
#include <pthread.h>
#include <errno.h>
#include <termios.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/timeb.h>
#include <sys/types.h>
//#include <libusb-1.0/libusb.h>
#include <unistd.h>
#include <math.h>


#define TRUE   1
#define FALSE  0
typedef unsigned char gbool;
//~----------------------------------------------------------------------------
int main(int argc, char *argv[]) {

  int fin,status,numberbytes;
  char fname[80]="m6_1.bin";
  unsigned short buf[0x14000];
  int i,j,k,nturn,nt=0,nsectors=0;
  double sum[18]={18*0},sector[18],Srpm;
  double x,sum2[18]={18*0},sigma[18],sumall=0,sectorM;
  
  //	printf("speed=");
  //	scanf("%d",&Srpm);
  //        Srpm = 6000;
  //	printf("%d\n",Srpm);
  //	sector0=1000000./12./(Srpm/60.);

       	printf("file=");
       	scanf("%s",fname);

	printf("open %s\n",fname);

	fin = open(fname,O_RDONLY);
	if (fin == -1) {
	  printf("open file error %s  error=%d\n",fname,fin);
	  return(1);
	}

	numberbytes = read(fin, (char *)buf, sizeof(buf));
	if (numberbytes == -1) printf("errno=%d\n",errno);
	printf("read %d\n", numberbytes);
	close(fin);

	for (k=0; (0x8000 & buf[k]) == 0; k++) ;
	printf("%d  %04x\n",k,buf[k]);
	nturn = (numberbytes/2-k)/18-1;
	printf("nturn=%d\n",nturn);

	for (j=0; j<nturn; j++) {
	  
	  for (i=0; i<18; i++){
	    nsectors++;
	    if ((0x8000 & buf[18*j+i+k]) != 0) nt++;
	    x = ((0x7fff & buf[18*j+i+k])+.5)/5.;
	    sum[i] += x;
	    sum2[i] += x*x;
	    sumall += x;
	  //printf("%04x ", buf[18*j+i+k]);
	  }
	  //printf("\n");
	}
	sectorM = sumall/nsectors;
	Srpm = 60.*1000000./(18*sectorM);
 
printf("Nturns=%d RPM=%7.1f sec ava=%6.2f nsectors=%8d\n",nt,Srpm,sectorM,nsectors);

	for (i=0; i<18; i++) {
	  sector[i] = sum[i]/nt;
	  sigma[i] = sqrt((sum2[i] - sum[i]*sum[i]/nt)/nt); 
	  // sector[i] = sum[i]/nturn;
	  // sigma[i] = sqrt((sum2[i] - sum[i]*sum[i]/nturn)/nturn); 
	  printf("sec %2d  %7.2f usec  %7.2f  %5.1f   %5.2f",
		 i,sector[i],sectorM-sector[i],
		 100.*(sectorM-sector[i])/sectorM,sigma[i]);
	  printf("\n");
	}
	printf("\n");

	return(0);
}
